pkg/cloud/vsphere/services/kubeclient/kubeclient.go

```diff
@@ -0,0 +1,48 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubeclient
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	corev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/pkg/apis/cluster/v1alpha1"
+	remotev1 "sigs.k8s.io/cluster-api/pkg/controller/remote"
+	controllerClient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Context is used to get a Kubernetes client for a target cluster.
+type Context interface {
+	context.Context
+	GetCluster() *clusterv1.Cluster
+	GetControllerClient() controllerClient.Client
+}
+
+// New returns a new client for the target cluster using the KubeConfig
+// secret stored in the management cluster.
```

Comment:
this looks like a general library to get kubernetes core client from context, so it could be not specific to `target cluster` or `management cluster`?

Reply:
Hi @figo,

It’s specific to target cluster (per the godocs), and it does not get it from the context.

Reply:
Hi @figo,

Sorry, was on a walk when I responded. The client is formed by using the `remotev1` package from CAPI 0.1.4, which pulls the credentials and endpoint for the target cluster from a Secret stored on the management cluster (bootstrap or permanent) that contains the KubeConfig for a target cluster.

Reply:
Sorry, it is base on context not from, which godocs are you referring to? I just could not see this code that specific to Target cluster

Reply:
Hi @figo,

The ones on line 36:

> // New returns a new client for the target cluster using the KubeConfig
> // secret stored in the management cluster.

Reply:
Hi @figo,

By the way:

> so it could be not specific to target cluster or management cluster

It occurs to me that you may be making a distinction here that I do not. In this case, the initial target cluster which becomes the management cluster post-pivot is *still* a **target cluster** pre-pivot.